File: task-service/src/main/java/com/microtodo/task_service/repository/TaskRepository.java
package com.microtodo.task_service.repository;

import java.util.List;
import java.util.Optional;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import com.microtodo.task_service.model.Task;

@Repository
public interface TaskRepository extends JpaRepository<Task, Long> {
    List<Task> findByUserId(Long userId);
    List<Task> findByUserIdAndStatus(Long userId, Task.TaskStatus status);
    Optional<Task> findByIdAndUserId(Long id, Long userId);
    long countByUserId(Long userId);
}

File: android-app/app/src/main/java/com/microtodo/android/domain/repository/TaskRepository.kt
package com.microtodo.android.domain.repository

import com.microtodo.android.data.remote.dto.TaskStatus
import com.microtodo.android.domain.model.Task

/**
 * Repository interface for task operations
 * Domain layer defines the contract, data layer implements it
 */
interface TaskRepository {

    /**
     * Create a new task
     */
    suspend fun createTask(
        title: String,
        description: String?,
        userId: Long,
        status: TaskStatus? = null
    ): Result<Task>

    /**
     * Get all tasks, optionally filtered by user ID
     */
    suspend fun getAllTasks(userId: Long? = null): Result<List<Task>>

    /**
     * Get a specific task by ID
     */
    suspend fun getTaskById(id: Long): Result<Task>

    /**
     * Update an existing task
     */
    suspend fun updateTask(
        id: Long,
        title: String,
        description: String?,
        userId: Long,
        status: TaskStatus?
    ): Result<Task>

    /**
     * Delete a task
     */
    suspend fun deleteTask(id: Long): Result<Unit>
}
